### ISSUE TITLE
Segmentation fault on loss.backward

### ISSUE THREAD

**Issue by Omroth** [pytorch/pytorch#50986]
## 🐛 Bug

I’m getting a segmentation fault when running loss backward.

"Abnormal program termination: received signal 11 (Segmentation fault)"

## To Reproduce

int n_a = 31;
int n_b = 100;

torch::Tensor dynamic_parameters = torch::full({1}, /*value=*/0.0002, torch::dtype(torch::kFloat64).requires_grad(true));
torch::optim::SGD optimizer({dynamic_parameters}, /*lr=*/0.01);

torch::Tensor target = torch::zeros({n_a}, torch::dtype(torch::kFloat64).requires_grad(false));

torch::Tensor dots = torch::zeros({n_a, n_b}, torch::dtype(torch::kFloat64).requires_grad(false));
// I fill this with some data
torch::Tensor magnitudes = torch::zeros({n_b}, torch::dtype(torch::kFloat64).requires_grad(false));
// I fill out this with some data

torch::Tensor prediction = torch::matmul(dots, magnitudes);

torch::Tensor loss = torch::nn::functional::mse_loss(prediction, target);
loss.backward();

Here is the stack trace:

[New Thread 0x7fff00067700 (LWP 18408)]
[New Thread 0x7fff00868700 (LWP 18409)]

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
_PyMethodDef_RawFastCallDict () at /tmp/build/80754af9/python_1565725737370/work/Objects/call.c:464
464	/tmp/build/80754af9/python_1565725737370/work/Objects/call.c: No such file or directory.
(gdb) bt
#0  _PyMethodDef_RawFastCallDict () at /tmp/build/80754af9/python_1565725737370/work/Objects/call.c:464
#1  0x00005555556897c1 in _PyCFunction_FastCallDict (func=0x7fff39e0dbe0, args=<optimised out>, nargs=<optimised out>, kwargs=<optimised out>)
    at /tmp/build/80754af9/python_1565725737370/work/Objects/call.c:586
#2  0x00007ffef71602c3 in pybind11::detail::simple_collector<(pybind11::return_value_policy)1>::call (ptr=<optimised out>, this=0x7fffffffc888)
    at /home/ian/anaconda3/lib/python3.7/site-packages/pybind11/include/pybind11/cast.h:2032
#3  pybind11::detail::object_api<pybind11::detail::accessor<pybind11::detail::accessor_policies::str_attr> >::operator()<(pybind11::return_value_policy)1, pybind11::tuple&> (this=this@entry=0x7fffffffc940)
    at /home/ian/anaconda3/lib/python3.7/site-packages/pybind11/include/pybind11/cast.h:2187
#4  0x00007ffef71684a6 in pybind11::detail::print (args=..., kwargs=...)
    at /home/ian/anaconda3/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:2008
#5  0x00007ffef7168d3b in pybind11::print<(pybind11::return_value_policy)1, char const (&) [2]> ()
    at /home/ian/anaconda3/lib/python3.7/site-packages/pybind11/include/pybind11/pybind11.h:2037

## Environment

I'm building my application with gcc on ubuntu 18.04
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.14.0

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2070
Nvidia driver version: 450.102.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.13.3
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.2.89              hfd86e86_1  
[conda] mkl                       2019.4                      243  
[conda] mkl-include               2019.4                      243  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.14           py37ha843d7b_0  
[conda] mkl_random                1.1.0            py37hd6b4f25_0  
[conda] numpy                     1.18.4                   pypi_0    pypi
[conda] numpydoc                  0.9.1                      py_0  
[conda] pytorch                   1.7.1           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torch                     1.4.0                    pypi_0    pypi
[conda] torchaudio                0.7.2                      py37    pytorch
[conda] torchvision               0.8.2                py37_cu102    pytorch
[conda] torchviz                  0.0.1                    pypi_0    pypi

@ptrblck 

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @yf225 @glaringlee

**Comment by Omroth** [comment]
Changing

torch::Tensor dynamic_parameters = torch::full({1}, /*value=*/initial_guess, torch::dtype(torch::kFloat64).requires_grad(true));

to

torch::Tensor dynamic_parameters = torch::tensor({initial_guess}, torch::dtype(torch::kFloat64).requires_grad(true));

Stopped the crash.

**Comment by glaringlee** [comment]
@Omroth 
Can you update your reproduce example? 
With above example, I get the following error which is expected since no tensor in loss function need grad.
```
terminate called after throwing an instance of 'c10::Error'
  what():  element 0 of tensors does not require grad and does not have a grad_fn
```

**Comment by Omroth** [comment]
Well, I did fix that issue and it stopped crashing... so I guess something weird was going on with my original.

int n_a = 31;
    int n_b = 100;

    torch::Tensor dynamic_parameters = torch::full({1}, 0.0002, torch::dtype(torch::kFloat64).requires_grad(true));
    torch::optim::SGD optimizer({dynamic_parameters}, 0.01);

    torch::Tensor target = torch::zeros({n_a}, torch::dtype(torch::kFloat64).requires_grad(false));

    torch::Tensor dots = torch::zeros({n_a, n_b}, torch::dtype(torch::kFloat64).requires_grad(true));
    // I fill this with some data
    torch::Tensor magnitudes = torch::zeros({n_b}, torch::dtype(torch::kFloat64).requires_grad(true));
    // I fill out this with some data

    torch::Tensor prediction = torch::matmul(dots, magnitudes);

    pybind11::gil_scoped_release no_gil;

    torch::Tensor loss = torch::nn::functional::mse_loss(prediction, target);

    optimizer.zero_grad();

    loss.backward();

    optimizer.step();
